File: Java/src/main/java/com/gildedrose/GildedRose.java
package com.gildedrose;

class GildedRose {
    Item[] items;

    public GildedRose(Item[] items) {
        this.items = items;
    }

    public void updateQuality() {
      for (Item item : items)
          extracted(item);
    }

    private static void extracted(Item item) {
        String name = item.getName();
        if (!name.equals("Aged Brie")
            && !name.equals("Backstage passes to a TAFKAL80ETC concert")) {
          if (item.getQuality() > 0) {
            if (!name.equals("Sulfuras, Hand of Ragnaros")) {
              item.setQuality(item.getQuality() - 1);
            }
          }
        } else {
          if (item.getQuality() < 50) {
            item.setQuality(item.getQuality() + 1);

            if (name.equals("Backstage passes to a TAFKAL80ETC concert")) {
              if (item.getSellIn() < 11) {
                if (item.getQuality() < 50) {
                  item.setQuality(item.getQuality() + 1);
                }
              }

              if (item.getSellIn() < 6) {
                if (item.getQuality() < 50) {
                  item.setQuality(item.getQuality() + 1);
                }
              }
            }
          }
        }

        if (!name.equals("Sulfuras, Hand of Ragnaros")) {
          item.setSellIn(item.getSellIn() - 1);
        }

        if (item.getSellIn() < 0) {
          if (!name.equals("Aged Brie")) {
            if (!name.equals("Backstage passes to a TAFKAL80ETC concert")) {
              if (item.getQuality() > 0) {
                if (!name.equals("Sulfuras, Hand of Ragnaros")) {
                  item.setQuality(item.getQuality() - 1);
                }
              }
            } else {
              item.setQuality(item.getQuality() - item.getQuality());
            }
          } else {
            if (item.getQuality() < 50) {
              item.setQuality(item.getQuality() + 1);
            }
          }
        }
    }
}

File: Java/src/main/java/com/gildedrose/GildedRose.kt
package com.gildedrose

import com.gildedrose.ItemType.*

class GildedRose(var items: Array<Item>) {
    fun updateQuality() {
        for (item in items) {
            update(item)
        }
    }
}

private fun update(item: Item) {
    val itemType = typeOf(item)
    age(itemType, item)
    degrade(itemType, item)
}

private fun degrade(itemType: ItemType, item: Item) {
    when (itemType) {
        BRIE -> {
            if (item.quality < 50) {
                item.quality = item.quality + 1
            }
            if (item.sellIn < 0) {
                if (item.quality < 50) {
                    item.quality = item.quality + 1
                }
            }
        }

        PASS -> {
            if (item.quality < 50) {
                item.quality = item.quality + 1

                if (item.sellIn < 10) {
                    if (item.quality < 50) {
                        item.quality = item.quality + 1
                    }
                }
                if (item.sellIn < 5) {
                    if (item.quality < 50) {
                        item.quality = item.quality + 1
                    }
                }
            }
            if (item.sellIn < 0) {
                item.quality = item.quality - item.quality
            }
        }


        SULFURAS -> {
        }

        OTHER -> {
            if (item.quality > 0) {
                item.quality = item.quality - 1
            }
            if (item.sellIn < 0) {
                if (item.quality > 0) {
                    item.quality = item.quality - 1
                }
            }
        }
    }
}

private fun age(itemType: ItemType, item: Item) {
    when (itemType) {
        SULFURAS -> {
        }

        else -> item.sellIn = item.sellIn - 1
    }
}

enum class ItemType {
    SULFURAS, PASS, BRIE, OTHER
}

fun typeOf(item: Item): ItemType = when (item.name) {
    "Sulfuras, Hand of Ragnaros" -> SULFURAS
    "Backstage passes to a TAFKAL80ETC concert" -> PASS
    "Aged Brie" -> BRIE
    else -> OTHER
}
